crates/pallet-domains/src/lib.rs

```diff
@@ -656,3 +764,36 @@ where
         }
     }
 }
+
+/// Calculates the new tx range based on the bundles produced during the interval.
+pub fn calculate_tx_range(
+    cur_tx_range: U256,
+    actual_bundle_count: u64,
+    expected_bundle_count: u64,
+) -> U256 {
+    if actual_bundle_count == 0 || expected_bundle_count == 0 {
+        return cur_tx_range;
+    }
+
+    let actual = U256::from(actual_bundle_count);
+    let expected = U256::from(expected_bundle_count);
+    let new_tx_range = actual.saturating_mul(&cur_tx_range);
+    let new_tx_range = if let Some(div) = new_tx_range.checked_div(&expected) {
+        div
+    } else {
+        return cur_tx_range;
+    };
+    let upper_bound = cur_tx_range.saturating_mul(&U256::from(4_u64));
+    let lower_bound = if let Some(div) = cur_tx_range.checked_div(&U256::from(4_u64)) {
+        div
+    } else {
+        return cur_tx_range;
+    };
+    if new_tx_range > upper_bound {
```

Comment:
You could follow solution_range logic here https://github.com/subspace/subspace/blob/f3157c3e1aab55115d68177a583f1c7b67afb1d6/crates/subspace-verification/src/lib.rs#LL303C6-L303C11